File: java-suite/src/main/java/de/fraunhofer/iem/collections/iterators/iterable/inCustomClass/FiveInts.java
package de.fraunhofer.iem.collections.iterators.iterable.inCustomClass;

import java.util.Iterator;
import java.util.Spliterator;
import java.util.function.Consumer;

/**
 * Demonstrates the custom class implements the Iterable in Java
 *
 * @author Ranjith Krishnamurthy
 */
public class FiveInts implements Iterable<Integer> {
    private static final int SIZE = 5;
    private final int[] data = new int[5];

    public FiveInts(int first, int second, int third, int fourth, int fifth) {
        this.data[0] = first;
        this.data[1] = second;
        this.data[2] = third;
        this.data[3] = fourth;
        this.data[4] = fifth;
    }

    @Override
    public Iterator<Integer> iterator() {
        return new FiveIntsIterator();
    }

    @Override
    public void forEach(Consumer<? super Integer> action) {
        Iterable.super.forEach(action);
    }

    @Override
    public Spliterator<Integer> spliterator() {
        return Iterable.super.spliterator();
    }

    private class FiveIntsIterator implements Iterator<Integer> {
        int currentCursor = -1;
        int nextCursor;

        @Override
        public boolean hasNext() {
            return currentCursor < SIZE - 1;
        }

        @Override
        public Integer next() {
            nextCursor = ++currentCursor;
            return data[nextCursor];
        }

        @Override
        public void remove() {
            Iterator.super.remove();
        }

        @Override
        public void forEachRemaining(Consumer<? super Integer> action) {
            Iterator.super.forEachRemaining(action);
        }
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/collections/iterators/iterable/inCustomClass/FiveInts.kt
package de.fraunhofer.iem.collections.iterators.iterable.inCustomClass

/**
 * Demonstrates the custom class implements the Iterable (read only) in Kotlin
 *
 * @author Ranjith Krishnamurthy
 */
class FiveInts(first: Int, second: Int, third: Int, fourth: Int, fifth: Int) : Iterable<Int> {
    private val data = IntArray(5)

    override fun iterator(): Iterator<Int> {
        return FiveIntsIterator()
    }

    private inner class FiveIntsIterator : Iterator<Int> {
        var currentCursor = -1
        var nextCursor = 0

        override fun hasNext(): Boolean {
            return currentCursor < SIZE - 1
        }

        override fun next(): Int {
            nextCursor = ++currentCursor
            return data[nextCursor]
        }
    }

    companion object {
        private const val SIZE = 5
    }

    init {
        data[0] = first
        data[1] = second
        data[2] = third
        data[3] = fourth
        data[4] = fifth
    }
}